Proposals/20200511-HVPA-design.md

```diff
@@ -0,0 +1,129 @@
+
+HVPA controller would calculate the total memory and CPU recommended by the recommenders (eg. HPA or VPA) by multiplying pod resources and the number of replicas, and would find the bucket for the calculated total resources. HVPA controller would then divide the bucket linearly to calculate the actual `replicas` and `resources` to which it will scale the application such that the total allocated resources to all its replicas is same as recommeded by the recommenders.
+
+### User Stories (optional)
+
+#### Story 1
+A user wants to scale both horizontally and vertically with folloing constraints:
+1. minimum replicas: 1, maximum replicas: 8
+1. minimum memory: 500M, maximum memory: 25G
+1. mimimum CPU: 0m, maximum CPU: 24000m
+
+The above requirement can be denoted as a single bucket.
+HVPA controller would divide the "CPU sub-bucket" and the "memory sub-bucket" linearly among the replica counts such that:
+1. each CPU bucket is of same size, ie, `3000m`
+1. each memory bucket is of same size, ie, `(10G - 100M) / 8`or `1.18G`
+
+Following is an image describing the divided CPU bucket. The numbers in the shaded boxes denote the maximum total resources that can be allocated for a replica count.
+![CPU Story 1](./images/CPUStory1.png)
+
+HVPA controller will take the total cpu and memory resources recommended by HPA or VPA, and will scale the application both horizontally and vertically such that:
+1. the total resoureces allocated to application is same as that recommended by HPA or VPA
+1. the total allocated resource is not more than the maximum total resource for that replica count. This way the resource per replica can be capped to a certain maximum value
+
+For example: If an application is running with 2 replicas with 6 cores each, and VPA and HPA recommendations are 12 cores and 4 replicas respectively, then HVPA would scale the application to 3 replicas and 8 cores. This is because both HPA and VPA recommend 24 cores, and this total lies in the 3rd column of 3 replicas. 
+
+#### Story 2
+A user wants to scale his/her application such that:
+1. It scales only vertically with 1 replica till it reaches 1 core and 2GB memory
+1. When more than 2GB memory is required, then the application should be scaled horizontally and vertically till memory reaches 20GB, and CPU reaches 9 cores, until maximum replica count of 8.
+
+In this case, user can provide 2 buckets:
+1. First bucket:
+   * Replicas: minimum: 1, maximum: 1
+   * CPU: minimum: 0, maximum: 1000m
+   * Memory: minimum: 0, maximum: 2G
+1. Second bucket:
+   * Replicas: minimum: 2, maximum: 8
+   * CPU: minimum: 1, maximum: 9000m
+   * Memory: minimum: 2G, maximum: 20G
+
+### Notes/Constraints/Caveats
+
+#### Different recommendations from HPA and VPA
+In such a case, higher recommendations will be used to calculate the desired replica count. This replica count would be used to calculate desired replica size.
```

Comment:
I am really not sure whether this will work out nicely in the end. HPA has no view onto/memory of the past / decay stats and acts immediately. VPA works differently. I am more and more sceptical that mixing them is a good idea. HPA will always be the jumpy guy here and when it comes to CPU, which is compressible, I would prefer less jumpiness. Their (HPA and VPA) algorithms differ quite considerably.

Reply:
agreed. We have an option to turn off HPA by setting `.spec.hpa.deploy` to `false`.  In this case, we will only be using VPA's recommendations.
This is also covered under the sub-heading `HPA domination` in this doc.